File: 2024/d07/src/main.py
import re


input = open("input").read()


def solve(i, acc, t, nums):
    # print(i, acc, t, nums)
    if i == len(nums):
        return acc == t

    return (
        solve(i + 1, acc + nums[i], t, nums)
        or solve(i + 1, acc * nums[i], t, nums)
        or solve(i + 1, int(str(acc) + str(nums[i])), t, nums)
    )


out = 0
for line in input.splitlines():
    first, rest = line.split(":")
    total = int(first)
    nums = [int(n) for n in re.findall(r"\d+", rest)]

    if solve(1, nums[0], total, nums):
        out += total

print(out)

# line = "3267: 81 40 27"
# first, rest = line.split(":")
# total = int(first)
# nums = [int(n) for n in re.findall(r"\d+", rest)]
#
# print(solve(1, nums[0], total, nums))

File: 2024/d07/src/main.rs
const INPUT: &str = include_str!("input");

fn solve(i: usize, acc: u64, target: u64, nums: &[u64]) -> bool {
    if i == nums.len() {
        return acc == target;
    }

    solve(i + 1, acc + nums[i], target, nums)
        || solve(i + 1, acc * nums[i], target, nums)
        || solve(
            i + 1,
            (acc.to_string() + &nums[i].to_string()).parse().unwrap(),
            target,
            nums,
        )
}

fn main() {
    let mut out = 0;
    for line in INPUT.lines() {
        let (first, rest) = line.split_once(":").unwrap();
        let total: u64 = first.parse().unwrap();
        let nums: Vec<u64> = rest
            .split_whitespace()
            .map(|n| n.parse().unwrap())
            .collect();

        if solve(1, nums[0], total, &nums) {
            out += total;
        }
    }

    println!("{out}");
}
